### PR TITLE
feat(redshift, rds): alarm on min/max connection count and long query duration

### LINKED ISSUE #232: [rds] Add ability to alarm on DB connections
### Feature scope

RDS

### Describe your suggested feature

Add ability to configure alarm on Cluster connections similar to `addCpuUsageAlarm`

### LINKED ISSUE #248: [redshift] Alarm for number of connections to Redshift cluster
### Feature scope

AWS/Redshift

### Describe your suggested feature

Alarm for maximum number of connections allowed to redshift cluster for connections widget created by .monitorRedshiftCluster. https://github.com/cdklabs/cdk-monitoring-constructs/blob/main/lib/monitoring/aws-redshift/RedshiftClusterMonitoring.ts#L81

### LINKED ISSUE #249: [redshift] Alarm on long query duration
### Feature scope

AWS/Redshift

### Describe your suggested feature

Allow for alarm on long query duration. https://github.com/cdklabs/cdk-monitoring-constructs/blob/main/lib/monitoring/aws-redshift/RedshiftClusterMonitoring.ts#L90

### PR DESCRIPTION
**Pull request by voho**
Adds alarm for min/max connection count and for long query duration.

Closes #248
Closes #249
Closes #232

---

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license_

### CODE REVIEW

File: lib/common/monitoring/alarms/DatabaseAlarmFactory.ts [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+export interface LowConnectionCountThreshold extends CustomAlarmThreshold {
+  readonly minConnectionCount: number;
+}
+
+export interface HighConnectionCountThreshold extends CustomAlarmThreshold {
+  readonly maxConnectionCount: number;
+}
+
+export class DatabaseAlarmFactory {
```
Comment (echeung-amzn) [inline lib/common/monitoring/alarms/DatabaseAlarmFactory.ts:17]: May be worth just making this generic towards the concept of connections. Not all databases have connection metrics (e.g. DynamoDB), and things with connections might not be databases (e.g. EC2 instances).

Comment (voho) [reply lib/common/monitoring/alarms/DatabaseAlarmFactory.ts:17]: Good callout, we could name it ConnectionAlarmFactory for now.

File: lib/common/monitoring/alarms/ConnectionAlarmFactory.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+      treatMissingData:
+        props.treatMissingDataOverride ?? TreatMissingData.MISSING,
+      comparisonOperator:
+        props.comparisonOperatorOverride ??
+        ComparisonOperator.LESS_THAN_THRESHOLD,
+      ...props,
+      disambiguator,
+      threshold: props.minConnectionCount,
+      alarmNameSuffix: "Connection-Count-Low",
+      alarmDescription: `Number of database connections is too low.`,
```
Comment (echeung-amzn) [inline lib/common/monitoring/alarms/ConnectionAlarmFactory.ts:39]:
```suggestion
      alarmDescription: `Number of connections is too low.`,
```

File: lib/common/monitoring/alarms/ConnectionAlarmFactory.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      treatMissingData:
+        props.treatMissingDataOverride ?? TreatMissingData.MISSING,
+      comparisonOperator:
+        props.comparisonOperatorOverride ??
+        ComparisonOperator.GREATER_THAN_THRESHOLD,
+      ...props,
+      disambiguator,
+      threshold: props.maxConnectionCount,
+      alarmNameSuffix: "Connection-Count-High",
+      alarmDescription: `Number of database connections is too high.`,
```
Comment (echeung-amzn) [inline lib/common/monitoring/alarms/ConnectionAlarmFactory.ts:58]:
```suggestion
      alarmDescription: `Number of connections is too high.`,
```

File: README.md [hunk 4]
```diff
@@ -92,8 +92,8 @@ You can browse the documentation at https://constructs.dev/packages/cdk-monitori
...
 | AWS Load Balancing (`.monitorNetworkLoadBalancer()`, `.monitorFargateApplicationLoadBalancer()`, `.monitorFargateNetworkLoadBalancer()`, `.monitorEc2ApplicationLoadBalancer()`, `.monitorEc2NetworkLoadBalancer()`) | System resources and task health | Unhealthy task count, running tasks count, (for Fargate/Ec2 apps) CPU/memory usage | Use for FargateService or Ec2Service backed by a NetworkLoadBalancer or ApplicationLoadBalancer |
 | AWS OpenSearch/Elasticsearch (`.monitorOpenSearchCluster()`, `.monitorElasticsearchCluster()`) | Indexing and search latency, disk/memory/CPU usage | Indexing and search latency, disk/memory/CPU usage, cluster status, KMS keys | |
-| AWS RDS (`.monitorRdsCluster()`) | Query duration, connections, latency, disk/CPU usage | Disk and CPU usage | |
-| AWS Redshift (`.monitorRedshiftCluster()`) | Query duration, connections, latency, disk/CPU usage | Disk and CPU usage | |
+| AWS RDS (`.monitorRdsCluster()`) | Query duration, connections, latency, disk/CPU usage | Connections, Disk and CPU usage | |
```
Comment (echeung-amzn) [inline README.md:95]:
```suggestion
| AWS RDS (`.monitorRdsCluster()`) | Query duration, connections, latency, disk/CPU usage | Connections, disk and CPU usage | |
```

File: README.md [hunk 5]
```diff
@@ -92,8 +92,8 @@ You can browse the documentation at https://constructs.dev/packages/cdk-monitori
...
+| AWS Redshift (`.monitorRedshiftCluster()`) | Query duration, connections, latency, disk/CPU usage | Query duration, Connections, disk and CPU usage | |
```
Comment (echeung-amzn) [inline README.md:96]:
```suggestion
| AWS Redshift (`.monitorRedshiftCluster()`) | Query duration, connections, latency, disk/CPU usage | Query duration, connections, disk and CPU usage | |
```

File: test/monitoring/aws-redshift/RedshiftClusterMonitoring.test.ts [hunk 6]
```diff
@@ -1,7 +1,7 @@
-import { Stack } from "aws-cdk-lib";
+import { Duration, Stack } from "aws-cdk-lib";
 import { Template } from "aws-cdk-lib/assertions";
 
-import { RdsClusterMonitoring } from "../../../lib";
+import { RedshiftClusterMonitoring } from "../../../lib";
```
Comment (echeung-amzn) [inline test/monitoring/aws-redshift/RedshiftClusterMonitoring.test.ts:4]: Yikes, good catch.

**Review by echeung-amzn — CHANGES_REQUESTED** [review]
Just some minor wording edits, but otherwise LGTM.

File: test/monitoring/aws-redshift/RedshiftClusterMonitoring.test.ts (hunk 6 above)
Comment (voho) [reply test/monitoring/aws-redshift/RedshiftClusterMonitoring.test.ts:4]: Glad as well!

**Review by echeung-amzn — APPROVED** [review]

**Pull request merged**